de.metas.adempiere.adempiere/base/src/main/java/de/metas/bpartner/service/impl/BPartnerDAO.java

```diff
@@ -1033,17 +1034,15 @@ public BPartnerLocationId retrieveBPartnerLocationId(@NonNull final BPartnerLoca
 			queryBuilder.orderByDescending(typeFilterColumnName); // "Y" first
 		}
 
-		final String orderByColumnName = getOrderByColumnNameForType(query.getType());
-		if (orderByColumnName != null)
-		{
-			queryBuilder.orderByDescending(orderByColumnName);
-		}
+		getOrderByColumnNameForType(query.getType()) // order by e.g. "IsDefaultShipToLocation"
+		.ifPresent(queryBuilder::orderByDescending);
+
 		queryBuilder.orderBy(I_C_BPartner_Location.COLUMNNAME_C_BPartner_Location_ID);
 
 		final I_C_BPartner_Location ownToLocation = queryBuilder
 				.create()
 				.first();
-		if (!query.isAlsoTryRelation() || ownToLocation != null)
+		if (!(query.isAlsoTryRelation() && query.getRelationBPartnerLocationId() != null) && ownToLocation != null)
```

Comment:
I think the javadoc of BPartnerLocationQuery is at odds with the boolean expression in this `if` (it's ok to change the javadoc too, but make it stop lying).
Maybe the code is too complicated not only for me but also for you?

Extracting its parts into more meaningful local variables might help..or, if that doesn'T work, adding some more comments.
Also, please take a look and update (or remove, if they aren't needed anymore) at the comments within this if-block.